server/src/main/java/org/elasticsearch/search/lookup/SourceLookup.java

```diff
@@ -24,27 +24,23 @@
 import org.elasticsearch.xcontent.XContentType;
 
 import java.io.IOException;
-import java.util.Collection;
+import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
 import java.util.function.Supplier;
 
-import static java.util.Collections.emptyMap;
-
-public class SourceLookup implements Map<String, Object> {
-
-    private LeafReader reader;
-    private CheckedBiConsumer<Integer, FieldsVisitor, IOException> fieldReader;
+public class SourceLookup {
```

Comment:
Can we double-check that removing the Map interface here doesn't break scripts that look things up via `_source['foo']`? There's a lot of dynamic typing that goes on in scripts and I'm a bit nervous that this will silently cause some issues.  Maybe @stu-elastic or @jdconrad have a view? If it works, then that's awesome as having it here has been annoying me for several years now :)

Reply:
I looked through this with @stu-elastic, and this looks like this should be a safe change. Everywhere we allow source access we do through a `DynamicMap` where we call the `source` method to get the source `Map` directly instead of using `SourceLookup`. The one edge case I can see is a custom plugin may have its own specially defined context; however, we don't guarantee any stable api right now.

Reply:
I think we need to also change the construction of the `asMap` member of `LeafSearchLookup`.  Although this hasn't caused any test failures, which either means we have a gap in test coverage or that this isn't actually used anymore.